src/app/cli/src/coda_change_snark_worker_test.ml

```diff
@@ -0,0 +1,102 @@
+  in
+  let snark_work_public_keys i =
+    if i = snark_worker_and_proposer_id then Some largest_public_key else None
+  in
+  let%bind testnet =
+    Coda_worker_testnet.test logger n proposers snark_work_public_keys
+      Cli_lib.Arg_type.Sequence ~max_concurrent_connections:None
+  in
+  let%bind new_block_pipe1, new_block_pipe2 =
+    let%map pipe =
+      Coda_worker_testnet.Api.validated_transitions_keyswaptest testnet
+        snark_worker_and_proposer_id
+    in
+    Pipe.fork ~pushback_uses:`Fast_consumer_only (Option.value_exn pipe).pipe
+  in
+  let wait_for_snark_worker_proof new_block_pipe public_key =
+    let found_snark_prover_ivar = Ivar.create () in
+    Pipe.iter new_block_pipe ~f:(fun {With_hash.hash= _; data= transition} ->
+        let completed_works =
+          Staged_ledger_diff.completed_works
+          @@ External_transition.staged_ledger_diff transition
+        in
+        if
+          List.exists completed_works
+            ~f:(fun {Transaction_snark_work.prover; _} ->
+              Logger.trace logger "Prover of completed work"
+                ~module_:__MODULE__ ~location:__LOC__
+                ~metadata:[("Prover", Public_key.Compressed.to_yojson prover)] ;
+              Public_key.Compressed.equal prover public_key )
+        then (
+          Logger.trace logger "Found snark prover ivar filled"
+            ~module_:__MODULE__ ~location:__LOC__
+            ~metadata:
+              [("public key", Public_key.Compressed.to_yojson public_key)] ;
+          Ivar.fill_if_empty found_snark_prover_ivar () )
+        else () ;
+        Deferred.unit )
+    |> don't_wait_for ;
+    Ivar.read found_snark_prover_ivar
+  in
+  Logger.trace logger "Waiting to get snark work from largest public key"
+    ~module_:__MODULE__ ~location:__LOC__
+    ~metadata:
+      [ ( "largest public key"
+        , Public_key.Compressed.to_yojson largest_public_key ) ] ;
+  let%bind () =
+    wait_for_snark_worker_proof new_block_pipe1 largest_public_key
+  in
+  let new_snark_worker =
+    List.find_map_exn Genesis_ledger.accounts ~f:(fun (_, account) ->
+        let public_key = Account.public_key account in
+        Option.some_if
+          (not @@ Public_key.Compressed.equal largest_public_key public_key)
+          public_key )
+  in
+  Logger.trace logger "Setting new snark worker key"
+    ~metadata:
+      [("new snark worker", Public_key.Compressed.to_yojson new_snark_worker)]
+    ~module_:__MODULE__ ~location:__LOC__ ;
+  let%bind () =
+    let%map opt =
+      Coda_worker_testnet.Api.replace_snark_worker_key testnet
+        snark_worker_and_proposer_id (Some new_snark_worker)
+    in
+    Option.value_exn opt
+  in
+  let%bind () = wait_for_snark_worker_proof new_block_pipe2 new_snark_worker in
+  Logger.trace logger "Finished waiting for snark work with updated key"
+    ~module_:__MODULE__ ~location:__LOC__ ;
+  (* Testing that nothing should break if the snark worker is set to None *)
```

Comment:
This triggers https://github.com/CodaProtocol/coda/issues/664 which is probably fine.